Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP1StateMachine.swift

```diff
@@ -0,0 +1,457 @@
+        }
+
+        typealias Action = HTTPConnectionPool.StateMachine.Action
+
+        private var connections: HTTP1Connections
+        private var failedConsecutiveConnectionAttempts: Int = 0
+
+        private var queue: RequestQueue
+        private var state: State = .running
+
+        init(idGenerator: Connection.ID.Generator, maximumConcurrentConnections: Int) {
+            self.connections = HTTP1Connections(
+                maximumConcurrentConnections: maximumConcurrentConnections,
+                generator: idGenerator
+            )
+
+            self.queue = RequestQueue()
+        }
+
+        // MARK: - Events -
+
+        mutating func executeRequest(_ request: Request) -> Action {
+            switch self.state {
+            case .running:
+                if let eventLoop = request.requiredEventLoop {
+                    return self.executeRequestOnRequiredEventLoop(request, eventLoop: eventLoop)
+                } else {
+                    return self.executeRequestOnPreferredEventLoop(request, eventLoop: request.preferredEventLoop)
+                }
+            case .shuttingDown, .shutDown:
+                // it is fairly unlikely that this condition is met, since the ConnectionPoolManager
+                // also fails new requests immediately, if it is shutting down. However there might
+                // be race conditions in which a request passes through a running connection pool
+                // manager, but hits a connection pool that is already shutting down.
+                //
+                // (Order in one lock does not guarantee order in the next lock!)
+                return .init(
+                    request: .failRequest(request, HTTPClientError.alreadyShutdown, cancelTimeout: false),
+                    connection: .none
+                )
+            }
+        }
+
+        private mutating func executeRequestOnPreferredEventLoop(_ request: Request, eventLoop: EventLoop) -> Action {
+            if let connection = self.connections.leaseConnection(onPreferred: eventLoop) {
+                return .init(
+                    request: .executeRequest(request, connection, cancelTimeout: false),
+                    connection: .cancelTimeoutTimer(connection.id)
+                )
+            }
+
+            // No matter what we do now, the request will need to wait!
+            self.queue.push(request)
+            let requestAction: StateMachine.RequestAction = .scheduleRequestTimeout(
+                for: request,
+                on: eventLoop
+            )
+
+            if !self.connections.canGrow {
+                // all connections are busy and there is no room for more connections, we need to wait!
+                return .init(request: requestAction, connection: .none)
+            }
+
+            // if we are not at max connections, we may want to create a new connection
+            if self.connections.startingGeneralPurposeConnections >= self.queue.count {
```

Comment:
Is this counting the wrong thing? This seems to risk starting a general purpose connection because we have lots of loop specific work, but without consulting what that loop specific work is.